File: server/src/main/java/com/tomato/bookstore/service/ReviewService.java
package com.tomato.bookstore.service;

import com.tomato.bookstore.dto.ReviewCreateDTO;
import com.tomato.bookstore.dto.ReviewDTO;
import java.util.List;

/** 书评服务接口 */
public interface ReviewService {
  /**
   * 获取指定商品的所有书评
   *
   * @param productId 商品 ID
   * @return 书评列表
   */
  List<ReviewDTO> getReviewsByProductId(Long productId);

  /**
   * 获取用户的所有书评
   *
   * @param userId 用户 ID
   * @return 书评列表
   */
  List<ReviewDTO> getReviewsByUserId(Long userId);

  /**
   * 创建书评
   *
   * @param productId 商品 ID
   * @param userId 用户 ID
   * @param reviewCreateDTO 书评创建 DTO
   * @return 创建的书评
   */
  ReviewDTO createReview(Long productId, Long userId, ReviewCreateDTO reviewCreateDTO);

  /**
   * 更新书评
   *
   * @param reviewId 书评 ID
   * @param userId 用户 ID
   * @param reviewCreateDTO 书评创建 DTO
   * @return 更新后的书评
   */
  ReviewDTO updateReview(Long reviewId, Long userId, ReviewCreateDTO reviewCreateDTO);

  /**
   * 管理员更新书评
   *
   * @param reviewId 书评 ID
   * @param reviewCreateDTO 书评创建 DTO
   * @return 更新后的书评
   */
  ReviewDTO updateReviewByAdmin(Long reviewId, ReviewCreateDTO reviewCreateDTO);

  /**
   * 删除书评
   *
   * @param reviewId 书评 ID
   * @param userId 用户 ID
   */
  void deleteReview(Long reviewId, Long userId);

  /**
   * 管理员删除书评
   *
   * @param reviewId 书评 ID
   */
  void deleteReviewByAdmin(Long reviewId);

  /**
   * 计算产品的平均评分
   *
   * @param productId 商品 ID
   * @return 平均评分
   */
  Double calculateProductRating(Long productId);
}

File: client/src/services/review-service.ts
import { REVIEW_MODULE } from "@/constants/apiPrefix";
import type { ApiResponse } from "@/types/api";
import type {
  Review,
  ReviewCreateParams,
  ReviewUpdateParams,
} from "@/types/review";
import apiClient from "@/utils/apiClient";

/**
 * 书评服务
 */
export default {
  /**
   * 获取指定商品的所有书评
   *
   * @param {number} productId 商品 ID
   * @returns {Promise<ApiResponse<Review[]>>} 书评列表
   */
  getProductReviews(productId: number): Promise<ApiResponse<Review[]>> {
    return apiClient.get(`${REVIEW_MODULE}/product/${productId}`);
  },

  /**
   * 获取当前用户的所有书评
   *
   * @returns {Promise<ApiResponse<Review[]>>} 书评列表
   */
  getUserReviews(): Promise<ApiResponse<Review[]>> {
    return apiClient.get(`${REVIEW_MODULE}/user`);
  },

  /**
   * 获取指定用户的所有书评（仅管理员）
   *
   * @param {number} userId 用户 ID
   * @returns {Promise<ApiResponse<Review[]>>} 书评列表
   */
  getUserReviewsByAdmin(userId: number): Promise<ApiResponse<Review[]>> {
    return apiClient.get(`${REVIEW_MODULE}/user/${userId}`);
  },

  /**
   * 获取所有书评（仅管理员）
   *
   * @returns {Promise<ApiResponse<Review[]>>} 书评列表
   */
  getAllReviews(): Promise<ApiResponse<Review[]>> {
    return apiClient.get(`${REVIEW_MODULE}/all`);
  },

  /**
   * 创建书评
   *
   * @param {number} productId 商品 ID
   * @param {ReviewCreateParams} params 书评参数
   * @returns {Promise<ApiResponse<Review>>} 创建的书评
   */
  createReview(
    productId: number,
    params: ReviewCreateParams,
  ): Promise<ApiResponse<Review>> {
    return apiClient.post(`${REVIEW_MODULE}/product/${productId}`, params);
  },

  /**
   * 更新书评
   *
   * @param {number} reviewId 书评 ID
   * @param {ReviewUpdateParams} params 更新参数
   * @returns {Promise<ApiResponse<Review>>} 更新后的书评
   */
  updateReview(
    reviewId: number,
    params: ReviewUpdateParams,
  ): Promise<ApiResponse<Review>> {
    return apiClient.put(`${REVIEW_MODULE}/${reviewId}`, params);
  },

  /**
   * 管理员更新书评
   *
   * @param {number} reviewId 书评 ID
   * @param {ReviewUpdateParams} params 更新参数
   * @returns {Promise<ApiResponse<Review>>} 更新后的书评
   */
  updateReviewByAdmin(
    reviewId: number,
    params: ReviewUpdateParams,
  ): Promise<ApiResponse<Review>> {
    return apiClient.put(`${REVIEW_MODULE}/admin/${reviewId}`, params);
  },

  /**
   * 删除书评
   *
   * @param {number} reviewId 书评 ID
   * @returns {Promise<ApiResponse<string>>} 删除结果
   */
  deleteReview(reviewId: number): Promise<ApiResponse<string>> {
    return apiClient.delete(`${REVIEW_MODULE}/${reviewId}`);
  },

  /**
   * 管理员删除书评
   *
   * @param {number} reviewId 书评 ID
   * @returns {Promise<ApiResponse<string>>} 删除结果
   */
  deleteReviewByAdmin(reviewId: number): Promise<ApiResponse<string>> {
    return apiClient.delete(`${REVIEW_MODULE}/admin/${reviewId}`);
  },
};
